File: task4_python/SurvivalGame.py
""""
		    /*
		    * CSCI3180 Principles of Programming Languages
		    *
		    * --- Declaration ---
		    *
		    * I declare that the assignment here submitted is original except for source
		    * material explicitly acknowledged. I also acknowledge that I am aware of
		    * University policy and regulations on honesty in academic work, and of the
		    * disciplinary guidelines and procedures applicable to breaches of such policy
		    * and regulations, as contained in the website
		    * http://www.cuhk.edu.hk/policy/academichonesty/
		    *
		    * Assignment 2
		    * Name : Nigel Nicholas
		    * Student ID : 1155088791
		    * Email Addr : nigel7@cse.cuhk.edu.hk
		    */

"""

import random
import sys
from Player import Player
from Weapon import Weapon
from Obstacle import Obstacle
from Pos import Pos
from Human import Human
from Chark import Chark
from Rifle import Rifle
from Axe import Axe


class SurvivalGame(object):
	#initialize as class attribute
	n=0 
	Dimension=10
	Obstacle=2

	def __init__(self):
		print "Welcome to Kafustrok. Light blesses you."
		SurvivalGame.n = input("Input number of players: (an even number) ")

		self.teleportObjects= [0 for x in range(SurvivalGame.n + SurvivalGame.Obstacle)]
		

		#should create n/2 humans

		for i in range(SurvivalGame.n /2):
			self.teleportObjects[i] = Human(0,0,i,self)
			self.teleportObjects[i + SurvivalGame.n/2]= Chark(0,0,i,self)

		for i in range(SurvivalGame.Obstacle):
			self.teleportObjects[i+SurvivalGame.n] = Obstacle(0,0,i,self)

	
	def printBoard(self):
		#initialize to spaces
		printObject = [["  "for x in range(SurvivalGame.Dimension)] for y in range(SurvivalGame.Dimension)]

		#getting player:
		 

		for i in range(SurvivalGame.n):
			pos = self.teleportObjects[i].pos
			printObject[pos.x][pos.y]= self.teleportObjects[i].myString

		for i in range(SurvivalGame.n , SurvivalGame.n + SurvivalGame.Obstacle):
			pos = self.teleportObjects[i].pos
			printObject[pos.x][pos.y]= "O" + str(i - SurvivalGame.n)

		sys.stdout.write(" ")
		for i in range(SurvivalGame.Dimension):
			sys.stdout.write("| %d  " % i)
		sys.stdout.write("|")
		sys.stdout.flush()
		print(" ")

		print "-"*55

		for row in range(SurvivalGame.Dimension):
			sys.stdout.write("%d" % row)
			for col in range(SurvivalGame.Dimension):
				sys.stdout.write("| %s "%printObject[row][col])
			sys.stdout.write("|")
			sys.stdout.flush()
			print(" ")
			print"-"*55

	def positionOccupied(self,randx,randy):

		for obj in self.teleportObjects:
			if isinstance(obj,Player):
				pos = obj.pos
				if pos.x == randx and pos.y == randy:
					return True
			else:
				pos = obj.pos
				if pos.x == randx and pos.y == randy:
					return True
		return False


	def getPlayer(self,randx,randy):#return Player:
		for obj in self.teleportObjects:
			if isinstance(obj,Player):
				pos = obj.pos
				if pos.x == randx and pos.y == randy:
					return obj

		return None

	def gameStart(self):
		turn=0
		numOfAlivePlayers = SurvivalGame.n
		while numOfAlivePlayers > 1:
			if turn == 0:
				for obj in self.teleportObjects:
					if isinstance(obj,Human):
						obj.teleport()
					elif isinstance(obj, Chark):
						obj.teleport()
					elif isinstance(obj, Obstacle):
						obj.teleport()
				print("Everything gets teleported..")
			self.printBoard()
			t = self.teleportObjects[turn]
			if t.health > 0:
				t.askForMove()
				print(" ")
			turn = (turn + 1) % SurvivalGame.n

			numOfAlivePlayers = 0

			for i in range(SurvivalGame.n):
				if self.teleportObjects[i].health > 0:
					numOfAlivePlayers+=1
		print"Game over."
		self.printBoard()
			








game=SurvivalGame()
game.gameStart()

File: task4_java/SurvivalGame.java

    /*
    * CSCI3180 Principles of Programming Languages
    *
    * --- Declaration ---
    *
    * I declare that the assignment here submitted is original except for source
    * material explicitly acknowledged. I also acknowledge that I am aware of
    * University policy and regulations on honesty in academic work, and of the
    * disciplinary guidelines and procedures applicable to breaches of such policy
    * and regulations, as contained in the website
    * http://www.cuhk.edu.hk/policy/academichonesty/
    *
    * Assignment 2
    * Name : Nigel Nicholas
    * Student ID : 1155088791
    * Email Addr : nigel7@cse.cuhk.edu.hk
    */
    

import java.util.Scanner;

public class SurvivalGame {
	private int n; // Number of player
	public final int D = 10; // dimension of board
	private final int O = 2; // Number of obstacles

	
	private Object[] teleportObjects;

	public static Scanner reader = new Scanner(System.in);

	public void printBoard() {
		String printObject[][] = new String[D][D];

		// init printObject
		for (int i = 0; i < D; i++)
			for (int j = 0; j < D; j++)
				printObject[i][j] = "  ";

		for (int i = 0; i < n; i++) {
			Pos pos = ((Player)teleportObjects[i]).getPos();
			printObject[pos.getX()][pos.getY()] = ((Player) teleportObjects[i]).getName();
		}

		for (int i = n; i < n+O; i++) {
			Pos pos = ((Obstacle)teleportObjects[i]).getPos();
			printObject[pos.getX()][pos.getY()] = "O" + Integer.toString(i-n);
		}

		// printing
		System.out.print(" ");
		for (int i = 0; i < D; i++)
			System.out.print(String.format("| %d  ", i));

		System.out.println("|");

		for (int i = 0; i < D * 5.5; i++)
			System.out.print("-");
		System.out.println("");

		for (int row = 0; row < D; row++) {
			System.out.print(row);
			for (int col = 0; col < D; col++)
				System.out.print(String.format("| %s ",
				printObject[row][col]));
			System.out.println("|");
			for (int i = 0; i < D * 5.5; i++)
				System.out.print("-");
			System.out.println("");
		}

	}

	public boolean positionOccupied(int randx, int randy) {

		for (Object o : teleportObjects) {
			if (o instanceof Player) {
				Pos pos = ((Player) o).getPos();
				if (pos.getX() == randx && pos.getY() == randy)
					return true;
			} else {
				Pos pos = ((Obstacle) o).getPos();
				if (pos.getX() == randx && pos.getY() == randy)
					return true;
			}

		}

		return false;
	}

	public Player getPlayer(int randx, int randy) {
		// TODO Auto-generated method stub
		for (Object o : teleportObjects) {
			if (o instanceof Player) {
				Pos pos = ((Player) o).getPos();
				if (pos.getX() == randx && pos.getY() == randy)
					return (Player) o;
			}
		}

		return null;
	}

	public int getN() {
		return n;
	}

	private  void init() {

		System.out.println("Welcome to Kafustrok. Light blesses you. \nInput number of players: (a even number)");
		n = reader.nextInt();

		teleportObjects = new Object[n + O];

		// create N/2 Humans

		for (int i = 0; i < n / 2; i++) {
			teleportObjects[i] = new Human(0, 0, i, this);
			teleportObjects[i + n / 2] = new Chark(0, 0, i, this);
		}

		// create O obstacles. You cannot stand there
		for (int i = 0; i < O; i++) {

			teleportObjects[i + n] = new Obstacle(0, 0, i, this);
		}

		// positions would be reinitialized later. 0,0 is dummy

	}

	private  void gameStart() {
		int turn = 0;
		int numOfAlivePlayers = n;
		while (numOfAlivePlayers > 1) {
			// teleport after every N turns
			if (turn == 0) {
				for (Object obj : teleportObjects) {
					if (obj instanceof Human)
						((Human) obj).teleport();
					else if (obj instanceof Chark)
						((Chark) obj).teleport();
					else if (obj instanceof Obstacle)
						((Obstacle) obj).teleport();
				}
				System.out.println("Everything gets teleported..");
			}
			printBoard();
			Player t = (Player) teleportObjects[turn];
			// t can move only if he is alive!
			if (t.health > 0) {
				// dynamic binding helps
				t.askForMove();
				System.out.println("\n");
				
			}
			turn = (turn + 1) % n;
			// count number of alive players
			numOfAlivePlayers = 0;

			for (int i = 0; i < n; i++) {
				if (((Player) teleportObjects[i]).health > 0)
					numOfAlivePlayers += 1;
			}

		}

		System.out.println("Game over.");
		printBoard();

	}

	public static void main(String[] args) {
		// TODO Auto-generated method stub
		// System.out.println(String.format("(%d,%d)", 3,4));
		SurvivalGame game = new SurvivalGame();
		game.init();
		game.gameStart();
	}
}
